### PR TITLE
unit tests added

### PR DESCRIPTION
**Pull request by ivan-kysil**

### CODE REVIEW

File: src/main/java/ikysil/training/dao/impl/InMemoryOrderDao.java [hunk 1]
```diff
@@ -13,30 +14,41 @@
...
-        return null;
+        storage.put(order.getId(), order);
+        return order;
     }
 
     @Override
     public OrderDto findOrder(String id) {
-        return null;
+        OrderDto result = null;
+        for (Map.Entry<String,OrderDto> entry : storage.entrySet()) {
```
Comment (ivan-kysil) [inline src/main/java/ikysil/training/dao/impl/InMemoryOrderDao.java]: can just 
```
return storage.get(id);
```

File: src/main/java/ikysil/training/dao/impl/InMemoryOrderDao.java [hunk 2]
```diff
@@ -13,30 +14,41 @@
...
     }
 
     @Override
     public List<OrderDto> findAll() {
-        return null;
+        List<OrderDto> everyDto = new ArrayList<>();
+        for (Map.Entry<String,OrderDto> entry : storage.entrySet()) {
+            everyDto.add(entry.getValue());
+        }
+        return everyDto;
```
Comment (ivan-kysil) [inline src/main/java/ikysil/training/dao/impl/InMemoryOrderDao.java]: can just
```
return new ArrayList<>(storage.values());
```

File: src/main/java/ikysil/training/ws/api/v1/dto/OrderDto.java [hunk 3]
```diff
@@ -22,4 +22,17 @@
...
+
+    public OrderDto(String id, String orderNumber, String sourceSystem, String userId, String userEmail) {
+        this.id = id;
+        this.orderNumber = orderNumber;
+        this.sourceSystem = sourceSystem;
+        this.userId = userId;
+        this.userEmail = userEmail;
+    }
+
+    public String getId() {
```
Comment (ivan-kysil) [inline src/main/java/ikysil/training/ws/api/v1/dto/OrderDto.java]: you already have all getters and setters, see lombok  annotation `@Data` on class.

File: src/test/java/ikysil/training/dao/OrderDaoTest.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        assertEquals(order1, testOrder1);
+
+        OrderDto testOrder2 = dao.findOrder("3257");
+        assertEquals(order2, testOrder2);
+
+        OrderDto newOrder = new OrderDto("3257","thirty three", "noneone", "2440", "newuser@gmail.com");
+        dao.updateOrder("3257", newOrder);
+
+        List<OrderDto> orders = dao.findAll();
+        System.out.println(orders);
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/dao/OrderDaoTest.java]: `System.out.println()` is only for local development purposes. 
it should not go to production.

File: src/test/java/ikysil/training/dao/OrderDaoTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        assertTrue(orders.contains(newOrder));
+    }
+
+
+    @After
+    public void after(){
+        dao.deleteOrder("1144");
+        dao.deleteOrder("3257");
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/dao/OrderDaoTest.java]: you do not need to clean after test as `OrderDaoTest` would be instantiated for each test and `new InMemoryOrderDao()` would be clean for each test.

File: src/test/java/ikysil/training/dao/OrderDaoTest.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    private OrderDto order2 = new OrderDto("3257", "twenty two", "none", "4024", "seconduser@gmail.com");
+
+    @Before
+    public void before(){
+        dao.saveOrder(order1);
+        dao.saveOrder(order2);
+    }
+
+    @Test
+    public void test(){
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/dao/OrderDaoTest.java]: you could add separate tests for . e.g.:
1 `saveOrder`
2 `updateOrder`, update should replace old value
3 `deleteOrder`
4 `findAll` works if empty and returns empty list.
5 test inserting null values
6 saving duplicates

File: src/test/java/ikysil/training/dao/OrderDaoTest.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import ikysil.training.ws.api.v1.dto.OrderDto;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.List;
+
+import static org.junit.Assert.*;
+
+public class OrderDaoTest {
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/dao/OrderDaoTest.java]: you testing here `InMemoryOrderDao` so test should be named `InMemoryOrderDaoTest`.
you can not test interface `OrderDao`.
This is common naming convention

File: src/test/java/ikysil/training/test.java [hunk 8]
```diff
@@ -0,0 +1,4 @@
+package ikysil.training;
+
+public class test {
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/test.java]: what is it? )

File: src/test/java/ikysil/training/service/OrderServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+import ikysil.training.ws.api.v1.dto.OrderDto;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.List;
+
+import static org.junit.Assert.*;
+
+public class OrderServiceTest {
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/service/OrderServiceTest.java]: `OrderServiceImplTest`

File: src/test/java/ikysil/training/service/OrderServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    private OrderDto order2 = new OrderDto("3257", "twenty two", "none", "4024", "seconduser@gmail.com");
+
+    @Before
+    public void before(){
+        serv.createOrder(order1);
+        serv.createOrder(order2);
+    }
+
+    @Test
+    public void test(){
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/service/OrderServiceTest.java]: fix the test

Comment (ivan-kysil) [inline src/test/java/ikysil/training/service/OrderServiceTest.java]: more scenario possible

File: src/test/java/ikysil/training/service/OrderServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        OrderDto testOrder2 = serv.getOrder("3257");
+        assertEquals(order2, testOrder2);
+    }
+
+    @Test
+    public void getOrdersTest(){
+        when(dao.findAll()).thenReturn(Arrays.asList(order1, order2)).thenReturn(new ArrayList<>());
+        List<OrderDto> orders = serv.getOrders();
+        assertFalse(orders.isEmpty());
+        assertEquals(orders.size(), 2);
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/service/OrderServiceImplTest.java]: jUnit assertions are fine, but I just want to show you more convenient assertion lib called `assertj`.
Add import for that 
`import static org.assertj.core.api.Assertions.assertThat;`
and then you can make it more readable and short:
```
assertThat(orders)
                .isNotEmpty()
                .hasSize(2)
                .containsExactly(order1, order2);
```

File: src/test/java/ikysil/training/service/OrderServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        orders = serv.getOrders();
+        assertTrue(orders.isEmpty());
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/service/OrderServiceImplTest.java]: try to replace all jUnit assertions with `assertj` assertions, they are more readable/short/flexible.
`assertj` assertions start with `assertThat(orders)...`

File: src/test/java/ikysil/training/service/OrderServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        OrderDto newOrder = new OrderDto("3257","thirty three", "noneone", "2440", "newuser@gmail.com");
+        serv.updateOrder("3257", newOrder);
+        serv.updateOrder("3257", newOrder);
+        serv.updateOrder("3257", newOrder);
+        verify(dao,times(3)).updateOrder(anyString(), any());
+    }
+
+    @Test
+    public void deleteOrderTest(){
+        doAnswer((Answer) invocation -> {
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/service/OrderServiceImplTest.java:73]: I guess you misuse `doAnswer` when use it to assert something inside. More over `dao` may be not called at all and test will path.
More reliable way to verify that `dao` was called with particular parameters is to 
```
verify(dao).deleteOrder("1144");
verify(dao).deleteOrder("3257");
```

File: src/test/java/ikysil/training/dao/InMemoryOrderDaoTest.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    @Test
+    public void dublicatesTest(){
+        dao.saveOrder(order1);
+        dao.saveOrder(order1);
+        dao.saveOrder(order2);
+        dao.saveOrder(order2);
+    }
+
+    @After
+    public void deleteOrderTest(){
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/dao/InMemoryOrderDaoTest.java]: weird name for @After method
I think you do not need @After here, as well better name may be `removeOrders`

File: src/test/java/ikysil/training/dao/InMemoryOrderDaoTest.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    public void insertNullTest(){
+        dao.saveOrder(null);
+    }
+
+    @Test
+    public void dublicatesTest(){
+        dao.saveOrder(order1);
+        dao.saveOrder(order1);
+        dao.saveOrder(order2);
+        dao.saveOrder(order2);
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/dao/InMemoryOrderDaoTest.java:63]: could you check in test something, e.g. that storage contains only 2 orders "order1" and "order2".

File: src/test/java/ikysil/training/service/OrderServiceImplTest.java (hunk 13 above)
Comment (StudentLgotnik) [reply src/test/java/ikysil/training/service/OrderServiceImplTest.java:73]: In mockito doc is written that doAnsver is one of way to test void method, and and i just decide to practice with writing Answer. If i run dao.deleteOrder() with another parameters, test will failed.

File: src/test/java/ikysil/training/dao/InMemoryOrderDaoTest.java [hunk 16]
```diff
@@ -19,35 +19,35 @@
...
-        assertEquals(dao.findOrder("3257"), newOrder);
+        assertThat(dao.findOrder("3257")).isEqualTo(newOrder);
     }
 
     @Test
     public void findAllTest(){
         List<OrderDto> orders = dao.findAll();
-        assertEquals(2, orders.size());
-        deleteOrderTest();
+        assertThat(orders.size()).isEqualTo(2);
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/dao/InMemoryOrderDaoTest.java]: assertJ has this kind of assertion for lists:
```
assertThat(orders).hasSize(2)
```
agree it is more human readable

File: src/main/java/ikysil/training/dao/impl/InMemoryOrderDao.java [hunk 17]
```diff
@@ -13,30 +14,36 @@
 @Component
```
Comment (ivan-kysil) [inline src/main/java/ikysil/training/dao/impl/InMemoryOrderDao.java:12]: could you remove TODOs in code where they are irrelevant

File: src/test/java/ikysil/training/service/OrderServiceImplTest.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+            Object[] objects = invocation.getArguments();
+            assertThat(objects[0].equals("1144") || objects[0].equals("3257")).isTrue();
+//            assertTrue(objects[0].equals("1144") || objects[0].equals("3257"));
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/service/OrderServiceImplTest.java]: commented code is an anti-pattern, remove it

File: src/test/java/ikysil/training/service/OrderServiceImplTest.java [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+        serv.updateOrder("3257", newOrder);
+        serv.updateOrder("3257", newOrder);
+        verify(dao,times(3)).updateOrder(anyString(), any());
+    }
+
+    @Test
+    public void deleteOrderTest(){
+        doAnswer((Answer) invocation -> {
+            Object[] objects = invocation.getArguments();
+            assertThat(objects[0].equals("1144") || objects[0].equals("3257")).isTrue();
```
Comment (ivan-kysil) [inline src/test/java/ikysil/training/service/OrderServiceImplTest.java:75]: better way is 
`assertThat(objects[0]).isIn("1144", "3257");`
how beautiful it is, you could not do it as easy with JUnit assertions.

**Review by StudentLgotnik — APPROVED** [review]

**Pull request merged**